code/game/machinery/suit_storage_unit.dm

```diff
@@ -516,7 +584,112 @@
 			span_notice("You escape the cramped confines of [src]!"))
 		open_machine()
 
+/obj/machinery/suit_storage_unit/multitool_act(mob/living/user, obj/item/tool)
+	if(state_open || !card_reader_installed || !panel_open)
+		return TRUE
+
+	if(locked)
+		balloon_alert(user, "unlock first!")
+		return
+
+	access_locked = !access_locked
+	balloon_alert(user, "access panel [access_locked ? "locked" : "unlocked"]")
+	return TRUE
+
 /obj/machinery/suit_storage_unit/attackby(obj/item/I, mob/user, params)
+	. = TRUE
+	var/obj/item/card/id/id = null
+	if(!card_reader_installed && panel_open && !state_open && is_operational && istype(I, /obj/item/stock_parts/card_reader))
+		if(locked)
+			balloon_alert(user, "unlock first!")
+			return
+
+		user.visible_message(span_notice("[user] is installing a card reader."),
+					span_notice("You begin installing the card reader."))
+		if(!do_after(user, 4 SECONDS, target = src))
+			return
+		if(!(!locked && !state_open && panel_open && is_operational && !card_reader_installed) || !user.transferItemToLoc(I, src))
+			return
+
+		card_reader_installed = TRUE
+		I.moveToNullspace()
+		qdel(I)
+
+		balloon_alert(user, "card reader installed")
```

Comment:
```suggestion
	. = TRUE
	var/obj/item/card/id/id = null

	if(istype(I, /obj/item/stock_parts/card_reader))
		if(card_reader_installed)
			balloon_alert(user, "card reader exists!")
			return

		if(locked)
			balloon_alert(user, "unlock first!")
			return

		if(!panel_open)
			balloon_alert(user, "panel is screwed shut!")
			return

		if(state_open)
			balloon_alert(user, "must be closed!")
			return

		user.visible_message(span_notice("[user] is installing a card reader."),
					span_notice("You begin installing the card reader."))

		if(!do_after(user, 4 SECONDS, target = src, extra_checks = CALLBACK(src, PROC_REF(check_interactable), user)))
			return
		if(!user.transferItemToLoc(I, src))
			return

		card_reader_installed = TRUE
		I.moveToNullspace()
		qdel(I)

		balloon_alert(user, "card reader installed")
```

Reply:
`if(!(!locked && !state_open && panel_open && is_operational && !card_reader_installed) || !user.transferItemToLoc(I, src))`

We need these checks after the do_after() proc. since it takes us 4 seconds during that short time someone else could along and do stuff to our unit so we need to double check again

```
if(card_reader_installed)
			balloon_alert(user, "card reader exists!")
```

We shoudn't do this because the suit storage unit has an auxiliary storage compartment, so we want the player to able to put the card reader in there if he wishes. short circuiting the if condition to immediately execute just when  
`if(istype(I, /obj/item/stock_parts/card_reader))` won't allow that storage code to execute.

It's the exact same when we want to put a screw driver in the auxiliary storage compartment when the unit is open but open it's panel when the unit is shut

Reply:
Take the multi tool for example, adding a balloon alert when the state is open is just not necessary since we have put the multitool in the storage compartment already

https://user-images.githubusercontent.com/110812394/236120186-ed9ada3e-0b2c-45e4-8a53-47b30bf9c640.mp4

Reply:
For the multitool, don't add a balloon alert when it's open since it gets transferred, but for the other checks it would be a good idea.

Reply:
I was not aware you could store items in the auxiliary storage compartment.  Just like the multitool, don't add a balloon alert when the item gets stored, but for everything else we should aim to have notifications why something isn't happening.

All those checks are pretty much covered in the `check_interactable` callback.  It can be attached to the `do_after()` proc as the `extra_checks` argument.

Reply:
`check_interactable()` has the exact opposite conditions which i require to install a card reader. Here is the proc
https://github.com/tgstation/tgstation/blob/9387db8634ea83ba77cb414f1c0130059d93b49d/code/game/machinery/suit_storage_unit.dm#L331-L341
And here are my conditions
`if(!card_reader_installed && panel_open && !state_open && is_operational && istype(I, /obj/item/stock_parts/card_reader))
.....`

Here are the differences
1. `check_interactable()` returns false if the machine is closed 
https://github.com/tgstation/tgstation/blob/9387db8634ea83ba77cb414f1c0130059d93b49d/code/game/machinery/suit_storage_unit.dm#L332-L333
But  I require the machine to be closed to be able to install the card reader so this would fail for me

2. `check_interactable()` returns false if the panel is open 
https://github.com/tgstation/tgstation/blob/9387db8634ea83ba77cb414f1c0130059d93b49d/code/game/machinery/suit_storage_unit.dm#L335-L336
 but i require the panel to be open and i don't have to worry about if the uv is enabled because you cannot open the panel if uv is enabled in the first place
https://github.com/tgstation/tgstation/blob/9387db8634ea83ba77cb414f1c0130059d93b49d/code/game/machinery/suit_storage_unit.dm#L583-L584

So for these reasons i cannot use `check_interactable()`. This proc is used to check if we can bring up the UI only, but we don't have any UI for installing card reader so there you go.

As for balloon alerts i fixed those already so it's all good

Reply:
Then I suggest you make a 2nd proc like `check_interactable` named something like `can_install_card_reader` that has the appropriate checks you are wanting.  This is improves readability and makes it easier to digest.

Reply:
Now that i can do. sure